File: ops/built-in/tbe/impl/dynamic/inv.py
# Copyright 2021 Huawei Technologies Co., Ltd
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
# http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ============================================================================
"""
inv
"""

import functools
import te.lang.cce as tbe
from te import tvm
from te.utils import para_check
from te.utils import shape_util
from te.lang.base.shape_classifier import classify
from te.lang.base.shape_classifier import Mode
import te.lang.base as tbe_base
from impl.util.platform_adapter import register_operator_compute
from impl.util.platform_adapter import register_operator

# define a scalar , value = 1
SCALAR_ONE = 1


# pylint: disable=locally-disabled,unused-argument
@register_operator_compute("Inv", op_mode="dynamic", support_fusion=False)
def inv_compute(input_x, output_y, kernel_name="inv"):
    """
    compute inv

    Parameters
    ----------
    input_x: TVM tensor
        the placeholder of input data
    output_y: TVM tensor
        the placeholder of output data
    kernel_name: str
        kernel name, default value is "inv"

    Returns
    -------
    res: TVM tensor
        the result of compute
    """
    dtype = input_x.dtype
    shape = shape_util.shape_to_list(input_x.shape)

    temp_const = tvm.const(SCALAR_ONE, dtype=dtype)
    temp_tensor = tbe.broadcast(temp_const, shape, dtype)
    res = tbe.vdiv(temp_tensor, input_x)

    return res


@para_check.check_op_params(para_check.REQUIRED_INPUT, para_check.REQUIRED_OUTPUT,
                            para_check.KERNEL_NAME)
@register_operator("Inv")
def inv(input_x, output_y, kernel_name="inv"):
    """
    algorithm: inv
    calculating data's reciprocal, y = 1 / x

    Parameters
    ----------
    input_x: dict
        shape and dtype of input, only support float16, float32, int32
    output_y: dict
        shape and dtype of output, should be same shape and type as input
    kernel_name: str
        kernel name, default value is "inv"

    Returns
    -------
    None
    """
    input_dtype = input_x.get("dtype").lower()
    check_list = ("float16", "float32", "int32")
    para_check.check_dtype(input_dtype, check_list, param_name = "input_x")

    schedules, tensors = [], []
    ins = classify([input_x], Mode.ELEWISE)
    for (_input_x,) in ins:
        with tbe_base.compute():
            x_shape = shape_util.variable_shape([_input_x])
            fuseshape = [1]
            fuseshape[0] = functools.reduce(lambda x, y: x * y, x_shape[0])
            data_input = tvm.placeholder(fuseshape, dtype = input_dtype,
                                         name = "data_input")
            res = inv_compute(data_input, output_y, kernel_name)
            tensors.append([data_input, res])
        with tvm.target.cce():
            sch = tbe.auto_schedule(res)
        schedules.append(sch)

    config = {"name": kernel_name,
              "tensor_list": tensors,
              "bool_storage_as_1bit": False}
    tbe.build(schedules, config)
    

File: ops/built-in/aicpu/impl/kernels/normalized/custom/research/inv.cc
/**
 * Copyright 2021 Huawei Technologies Co., Ltd
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

#include "inv.h"

#include "cpu_kernel_utils.h"
#include "utils/eigen_tensor.h"
#include "utils/kernel_util.h"

namespace {
const uint32_t kOutputNum = 1;
const uint32_t kInputNum = 1;
const char *kInv = "Inv";
constexpr int64_t kParallelDataNums = 128 * 1024;

#define INV_COMPUTE_CASE(DTYPE, TYPE, CTX)            \
  case (DTYPE): {                                     \
    uint32_t result = InvCompute<TYPE>(CTX);          \
    if (result != KERNEL_STATUS_OK) {                 \
      KERNEL_LOG_ERROR("Inv kernel compute failed."); \
      return result;                                  \
    }                                                 \
    break;                                            \
  }
}  // namespace

namespace aicpu {
uint32_t InvCpuKernel::Compute(CpuKernelContext &ctx) {
  // check params
  KERNEL_HANDLE_ERROR(NormalCheck(ctx, kInputNum, kOutputNum),
                      "[%s] check input and output failed.", kInv);
  auto data_type = ctx.Input(0)->GetDataType();
  switch (data_type) {
    INV_COMPUTE_CASE(DT_FLOAT16, Eigen::half, ctx)
    INV_COMPUTE_CASE(DT_FLOAT, float, ctx)
    INV_COMPUTE_CASE(DT_DOUBLE, double, ctx)
    INV_COMPUTE_CASE(DT_COMPLEX64, std::complex<float>, ctx)
    INV_COMPUTE_CASE(DT_COMPLEX128, std::complex<double>, ctx)
    default:
      KERNEL_LOG_ERROR("Inv kernel data type [%s] not support.",
                       DTypeStr(data_type).c_str());
      return KERNEL_STATUS_PARAM_INVALID;
  }
  return KERNEL_STATUS_OK;
}

template <typename T>
uint32_t InvCpuKernel::InvCompute(CpuKernelContext &ctx) {
  auto input_x = reinterpret_cast<T *>(ctx.Input(0)->GetData());
  auto output_y = reinterpret_cast<T *>(ctx.Output(0)->GetData());
  std::vector<int64_t> shape_x = ctx.Input(0)->GetTensorShape()->GetDimSizes();
  size_t shape_size = shape_x.size();
  size_t data_num = ctx.Input(0)->NumElements();    
  int64_t data_size = ctx.Input(0)->NumElements() * sizeof(T);
  if(shape_x.size() == 1){
    if (data_size <= kParallelDataNums){
      for (size_t i = 0; i < data_num; i++) {
          T x = T(*(input_x + i));
          if (x == static_cast<T>(0)) {
            *(output_y + i) = static_cast<T>(INFINITY);
          } else {
            *(output_y + i) = static_cast<T>(1.0) / x;
          }
      }
    }
    else{
      uint32_t min_core_num = 1;
      uint32_t max_core_num =
          std::max(min_core_num, aicpu::CpuKernelUtils::GetCPUNum(ctx) - 2);
      if (max_core_num > data_num) {
        max_core_num = data_num;
      }
      auto shard_inv = [&](size_t start, size_t end) {
        for (size_t i = start; i < end; i++){
          T x = T(*(input_x + i));
          if (x == static_cast<T>(0)) {
            *(output_y + i) = static_cast<T>(INFINITY);
          } else {
            *(output_y + i) = static_cast<T>(1.0) / x;
          }
        }
      };
      KERNEL_HANDLE_ERROR(
          CpuKernelUtils::ParallelFor(ctx, data_num, data_num / max_core_num,
                                      shard_inv),
          "Inv Compute failed.")
    }   
  }
  else{
    int64_t m = shape_x[shape_size - 2];
    int64_t n = shape_x[shape_size - 1];
    using MartixXd = Eigen::Matrix<T, Eigen::Dynamic, Eigen::Dynamic, Eigen::RowMajor>;
    if ( m * n > 0) {
      int64_t size_mn = m * n;    
      size_t martix_num = ctx.Input(0)->NumElements() / size_mn;

      if (data_size <= kParallelDataNums) {
        for (size_t i = 0; i < martix_num; i++) {
          Eigen::Map<MartixXd> martix_x(input_x + i * size_mn, m, n);
          Eigen::Map<MartixXd> martix_y(output_y + i * size_mn, m, n);
          martix_y = martix_x.cwiseInverse();
        }
      }      
      else {
        uint32_t min_core_num = 1;
        uint32_t max_core_num =
            std::max(min_core_num, aicpu::CpuKernelUtils::GetCPUNum(ctx) - 2);
        if (max_core_num > martix_num) {
          max_core_num = martix_num;
        }
        auto shard_inv = [&](size_t start, size_t end) {
          {
              for (size_t i = start; i < end; i++) {
                Eigen::Map<MartixXd> martix_x(input_x + i * size_mn, m, n);
                Eigen::Map<MartixXd> martix_y(output_y + i * size_mn, m, n);
                martix_y = martix_x.cwiseInverse();
              }
          }
        };
        KERNEL_HANDLE_ERROR(
            CpuKernelUtils::ParallelFor(ctx, martix_num,
                                        martix_num / max_core_num, shard_inv),
            "Inv Compute failed.")
      }      
     }
    }
  return KERNEL_STATUS_OK;
 }  
REGISTER_CPU_KERNEL(kInv, InvCpuKernel);
}  // namespace aicpu